### ISSUE TITLE
Problem with Database Loader

### ISSUE THREAD

**Issue by FranzBl** [dotnet/machinelearning-samples#722]
The Microsoft ML Database Loader reads data from database in Vector<T> format. If for example a database column has type "int", the delivered value is "Vector" type int etc. These data are impossible to train since values of this type are not supported. Feature conversion via OneHotEncoding doesn't help either.
What to do? How to get rid of these Vectors? Any suggestion?

Here's a code snippet illustrating the problem:

            DatabaseSource dbSource = new DatabaseSource(SqlClientFactory.Instance, connectionString, sqlCommand);
            DatabaseLoader loader = mlContext.Data.CreateDatabaseLoader<ModelInput>();

            IDataView dataView = loader.Load(dbSource);

            var pipeline = mlContext.Transforms.CopyColumns(outputColumnName: "Label", inputColumnName: "PROZENT")
                .Append(mlContext.Transforms.Categorical.OneHotEncoding(outputColumnName: "NAMEEncoded", inputColumnName: "NAME"))
                .Append(mlContext.Transforms.Categorical.OneHotEncoding(outputColumnName: "ZEITEncoded", inputColumnName: "ZEIT"))
                .Append(mlContext.Transforms.Concatenate("Features", "NAMEEncoded", "ZEITEncoded"))
                .Append(mlContext.Regression.Trainers.FastTree());
            var model = pipeline.Fit(dataView);

Running this code leads to an error:

```
System.ArgumentOutOfRangeException
  HResult=0x80131502
  Nachricht = Schema mismatch for label column '': expected Single, got Vector<Single>
Arg_ParamName_Name
  Quelle = Microsoft.ML.Data
  Stapelüberwachung:
   at Microsoft.ML.Trainers.TrainerEstimatorBase`2.CheckLabelCompatible(Column labelCol)
   at Microsoft.ML.Trainers.TrainerEstimatorBase`2.CheckInputSchema(SchemaShape inputSchema)
   at Microsoft.ML.Trainers.TrainerEstimatorBase`2.GetOutputSchema(SchemaShape inputSchema)
   at Microsoft.ML.Data.EstimatorChain`1.GetOutputSchema(SchemaShape inputSchema)
   at Microsoft.ML.Data.EstimatorChain`1.Fit(IDataView input)
   at HyperStockTrainingModel.Program.Train(MLContext mlContext, String dataPath) in Q:\Software\Apps\HyperStockTrainingModel\HyperStockTrainingModel\Program.cs:line 42
   at HyperStockTrainingModel.Program.Main(String[] args) in Q:\Software\Apps\HyperStockTrainingModel\HyperStockTrainingModel\Program.cs:line 23
```

**Comment by eerhardt** [comment]
@FranzBl - can you also show your `ModelInput` class? You may not have the correct attributes on it.

If you are using `LoadColumn` or `LoadColumnName` attribute on the `PROZENT` property, ensure you are only selecting a single column.

**Comment by FranzBl** [comment]
Here is the class declaration. The database records are read as Vector<int>, Vector<float> etc. Remark: Only NAME and ZEIT are used as features in the pipeline,

using Microsoft.ML.Data;

namespace HyperStockTrainingModel
{
    internal class ModelInput
    {
        [LoadColumn(0)]
        public int SCHLUESSEL { get; set; }

        [LoadColumn(1)]
        public string TITEL { get; set; }

        [LoadColumn(2)]
        public string NAME { get; set; }

        [LoadColumn(3)]
        public string ZEIT { get; set; }

        [LoadColumn(4)]
        public float WERT { get; set; }

        [LoadColumn(5)]
        public float PROZENT { get; set; }
    }


    public class ModelOutput
    {
        [ColumnName("Score")]
        public float PROZENT;
    }

}

**Comment by eerhardt** [comment]
@tannergooding - can you take a look? Is there a bug in the DatabaseLoader here?

**Comment by FranzBl** [comment]
Perhaps I should add that the same model training code works fine if I don't use the Database interface but export the data into a csv file and load this file instead. But of course this is acceptable for test purposes only. The program user shouldn't have to do that. In fact, he shouldn't need to know anything about databases at all, as well as about csv files. He wants to observe stock values and to be able to force the program to estimate how it could go on... So Database interface is mandatory, but I suppose this is still in preview status.

**Comment by CESARDELATORRE** [comment]
Just a check, I guess you are using the DatabaseLoader from Microsoft.ML.Experimental 0.16.0-preview, right?

Can you try by removing the LoadColumn(X) attributes and confirm if it behaves the same way if simply using the property name matching the column name in the D.B.?

Btw, in the next release (1.4.0 GA) to be published in around a week we're you can also use the LoadColumnName and ColumnName attributes when using the database.

You'll be able to do the following:

```
[LoadColumnName("MyColumnInDB")]      //Column name as defined in the D.B.
[ColumnName("ColumnNameInSchema")] //I'm renaming for the schema
public int MyPropertyName;                       //C# property name to be used
```

**Comment by CESARDELATORRE** [comment]
Also, can you try to use **float** instead of **int** for the first column named SCHLUESSEL? 

Usually, most algorithms require float columns, so using float, bool or string should be the by default values.

**Comment by tannergooding** [comment]
Should be resolved with https://github.com/dotnet/machinelearning/pull/4403

**Comment by luisquintanilla** [comment]
Hi @FranzBl 

We have recently submitted a PR to our docs that might help with this issue.

In short, ML.NET algorithms / transforms expect numerical data to be a single-precision floating-point value, or Single. In the DB this is represented as REAL. Therefore, assuming your `SCHLUESSEL` property is an `int` in the DB, you would load it as such.

```csharp
var sqlCommand = "SELECT CAST(SCHLUESSEL as REAL) as SCHLUESSEL, <...other columns> FROM <YOUR-DB>"
```

Your `ModelInput` class would be:

```csharp
internal class ModelInput
{
[LoadColumn(0)]
public float SCHLUESSEL { get; set; }

//...other properties
}
```

The changes are not live on the Microsoft Docs site yet but you can see the doc here.

https://github.com/dotnet/docs/blob/master/docs/machine-learning/how-to-guides/load-data-ml-net.md#load-data-from-a-relational-database

**Comment by FranzBl** [comment]
I suppose the int value should not be the reason, since it is not used as a feature.

**Comment by FranzBl** [comment]
Schluessel is the primary key for the database table, a counter from 1 to ...

**Comment by tannergooding** [comment]
It's not the reason. There was a bug which I've fixed in the PR listed above.

I believe @luisquintanilla and @CESARDELATORRE are just trying to help share some "best practices" information

**Comment by FranzBl** [comment]
> Should be resolved with [dotnet/machinelearning#4403](https://github.com/dotnet/machinelearning/pull/4403)

Will be soon an update on nuget.org? I wouldn't like to reinstall the package from source, since I work on a project in Visual Studio, and there packages are most comfortably updated via nuget.

**Comment by FranzBl** [comment]
Another issue in this context, probably a design problem: DatabaseLoader requires its own database connection:
            DatabaseSource dbSource = new DatabaseSource(SqlClientFactory.Instance, connectionString, sqlCommand);
            DatabaseLoader loader = mlContext.Data.CreateDatabaseLoader<ModelInput>();
I suppose the idea behind this is that software developers will train their models in standalone programs that are not delivered with their software, to which only the trained models are passed.
I would like to integrate model training in a larger Windows forms solution. There the model will not be static, but the training data will grow with time, when new values are detected and stored in the database. So a regular training of the model will be required to enhance performance.
In this Windows program a SqlConnection is open all the time, from beginning to the end. So it would be preferable if DatabaseSource would have an additional constructor to pass an existing SqlConnection.

**Comment by FranzBl** [comment]
> Should be resolved with [dotnet/machinelearning#4403](https://github.com/dotnet/machinelearning/pull/4403)

I would like to test the bug fix, but if I simply download the corrected version of DatabaseSource I can't compile it since there are security protected classes inside. Is there any way I could help testing, or should I wait for a new version of Microsoft.ML.Experimental on nuget.org?

**Comment by tannergooding** [comment]
> Will be soon an update on nuget.org? I wouldn't like to reinstall the package from source, since I work on a project in Visual Studio, and there packages are most comfortably updated via nuget.

The PR is yet to be merged (it should happen today once CI completes). @eerhardt might be able to share how you can best try it out.

**Comment by CESARDELATORRE** [comment]
@FranzBl - I expect this fix will be merged in master asap. 

Recently, we're now publishing daily-build-drops on the following Nuget feed, so as soon as the package with the fix is published as preview-daily drop, you should be able to test it from the feed:

https://pkgs.dev.azure.com/dnceng/public/_packaging/MachineLearning/nuget/v3/index.json

Confirm you have access to the feed, ok? 👍

**Comment by FranzBl** [comment]
Excuse me, boys, I'm not in a hurry, simply I wanted to ask whether I could help with testing, since I was the one who found the bug.
Do that what has to be done, and do it thoroughly. I may wait for the next version.
I was long enough in software business to know what it means to go from a bug fix to a new tested version... :-)
